### DISCUSSION TITLE
StartupHooks and EnvironmentVariables in .NET (not ASP.NET)

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/runtime#65007] [Ideas] [answered]
I'm struggling a bit with startup hooks and environment variables in runtimeConfig.json.
Does this work in normal .NET? The reason is that I don't want to (and cannot) set global variables.
I'm especially interested in DOTNET_STARTUP_HOOKS and DOTNET_STARTUP_HOOKS.

Obviously the current runtimeConfig.json does not support to set environment variables at startup which I think would solves a lot of my problems in general. The docs and the schema do not say anything about this.

**Comment by vitek-karas** [comment, marked as answer]
You can set a runtime property `STARTUP_HOOKS` in the `.runtimeconfig.json`. The ENV `DOTNET_STARTUP_HOOKS` sets this property effectively (if both are present, they will be concatenated with the ENV going first).

Curious: What use case do you have to set startup hooks in `.runtimeconfig.json`?

**Reply by msedi** [reply]
@vitek-karas . Thanks for the info. I need to try. The sole reason why I need this is to startup the runtime with DOTNET_Thread_UseAllCpuGroups=true.
This is currently not possible in the runtimeConfig json and obviously needs to be set before the runtime starts up. I could do this by a batch startup but that seems a bit clunky. Obviously the option has been removed from .NET because it was available in .NET Framework.

**Reply by vitek-karas** [reply]
I think it would reasonable to make a request to read this value from both env and runtime config. We already have other options which allow this.

Why do you need to do this in startup hook? If you own the application, running code at the begining of `Main` is the same as doing it in startup hook.

**Reply by msedi** [reply]
@vitek-karas: OK. Thank you. I will open up a request.

Regarding your question: If this is the case the you are truly right. I was thinking that the startup are up with the runtime. Maybe I'm wrong but I thought the application executable (assume: ConsoleApp.exe) is just a starter that starts up the runtime and then runs the code in the application assembly (ConsoleApp.dll). So I thought the runtime needs the necessary runtime environment variables (in my case: DOTNET_Thread_UseAllCpuGroups) before startup of the application executable.

If that can be done differently I agree, then I don't event need an extension of the runtimeConfig.json.

In general that question is though why the runtimeConfig.json in .NET behaves differently considering environment variables than in ASP.NET? At least the environment variables could be treated the same way.

**Reply by vitek-karas** [reply]
I don't know when is this ENV variable read (would have to dig into the sources) - my point was that startup hooks are managed code which executes right before `Main` - so they're really no different than anything running at the beginning of `Main`. So if you can make your scenario work with startup hooks, you can also make it work within `Main`.

You're right about the executable, it's just a small "shim" which finds and loads the runtime and tells it to run the managed assembly. But runtime tends to be as lazy as possible, so it will probably read the configuration value only at the point where it really needs it for the first time.

**Reply by msedi** [reply]
@vitek-karas: Thanks for taking your time on this. If have added a request: #65457.
